src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -859,88 +847,113 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
-    int64_t distance = dest - pc();                                                   \
-    if (is_simm21(distance) && ((distance % 2) == 0)) {                               \
-      Assembler::NAME(Rd, distance);                                                  \
-    } else {                                                                          \
-      assert_different_registers(Rd, temp);                                           \
-      int32_t offset = 0;                                                             \
-      movptr(temp, dest, offset);                                                     \
-      jalr(Rd, temp, offset);                                                         \
-    }                                                                                 \
-  }                                                                                   \
-  void MacroAssembler::NAME(Register Rd, Label &L, Register temp) {                   \
-    assert_different_registers(Rd, temp);                                             \
-    wrap_label(Rd, L, temp, &MacroAssembler::NAME);                                   \
+  // We can't patch C, i.e. if Label wasn't bound we need to patch this jump.
+  IncompressibleRegion ir(this);
+  if (is_simm21(distance) && ((distance % 2) == 0)) {
+    Assembler::jal(x0, distance);
+  } else {
+    assert(temp != noreg && temp != x0, "expecting a register");
+    int32_t offset = 0;
+    la(temp, dest, offset);
+    jr(temp, offset);
   }
+}
 
-  INSN(jal);
+void MacroAssembler::j(const Address &adr, Register temp) {
+  switch (adr.getMode()) {
+    case Address::literal: {
+      relocate(adr.rspec(), [&] {
+        j(adr.target(), temp);
+      });
+      break;
+    }
+    case Address::base_plus_offset: {
+      int32_t offset = ((int32_t)adr.offset() << 20) >> 20;
+      la(temp, Address(adr.base(), adr.offset() - offset));
+      jr(temp, offset);
+      break;
+    }
+    default:
+      ShouldNotReachHere();
+  }
+}
 
-#undef INSN
+void MacroAssembler::j(Label &lab, Register temp) {
+  assert_different_registers(x0, temp);
+  if (lab.is_bound()) {
+    MacroAssembler::j(target(lab), temp);
+  } else {
+    lab.add_patch_at(code(), locator());
+    MacroAssembler::j(pc(), temp);
+  }
+}
 
-#define INSN(NAME, REGISTER)                                       \
-  void MacroAssembler::NAME(Label &l, Register temp) {             \
-    jal(REGISTER, l, temp);                                        \
-  }                                                                \
+void MacroAssembler::jr(Register Rd, int32_t offset) {
+  assert(Rd != noreg, "expecting a register");
+  Assembler::jalr(x0, Rd, offset);
+}
 
-  INSN(j,   x0);
-  INSN(jal, x1);
+void MacroAssembler::call(const address dest, Register temp) {
+  assert_cond(dest != nullptr);
+  assert(temp != noreg, "expecting a register");
+  int32_t offset = 0;
+  la(temp, dest, offset);
+  Assembler::jalr(x1, temp, offset);
```

Comment:
Hi, Did you check the possible impact on performance of this change?

old: call() => mv(tmp, adr) => li(tmp, adr)
new: call() => la(tmp, adr) => auipc/movptr(tmp, adr)

So the new call() could emit a fixed-sized movptr depending on the `is_32bit_offset_from_codecache` check which should be slower than a li(tmp, adr).

Reply:
I checked that we use the use call with a code cache address for the places we care about such as:
I.e. JNI FastGetField and far_call()

But I now see I was mistaken regarding the interpreter, it seems like we don't have all the math implemented, so we do calls to SharedRuntime there, I'll fix those.

Thanks!

Reply:
Great, thanks.

I fixed it by using li() in la() when we wan't use auipc, running tests.

Reply:
That was bad idea.

Looking into something else.

Reply:
It's a bit annoying, since if we are about performance in the interpreter we should implement the math stubs, e.g. StubRoutines::dcos.

Reply:
I don't find a nice way to get li for them.
https://github.com/openjdk/jdk/blob/3b8227ba24c7bc05a8ea23801e3816e8fc80de4e/src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp#L272

Any suggestions?

As I said, if we want to be performant here we should implement the "StubRoutines::dpow()".

What you think?

Reply:
As I can't fix this with a few changes, and I don't want to mess up this PR.
I added check for li directly in the call.
Sanity tested with -XX:ReservedCodeCacheSize=2047M